pkg/controller/framework/shared_informer.go

```diff
@@ -206,128 +167,53 @@ func (s *sharedIndexInformer) HandleDeltas(obj interface{}) error {
 				if err := s.indexer.Update(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(updateNotification{oldObj: old, newObj: d.Object})
+				s.OnUpdate(old, d.Object)
 			} else {
 				if err := s.indexer.Add(d.Object); err != nil {
 					return err
 				}
-				s.processor.distribute(addNotification{newObj: d.Object})
+				s.OnAdd(d.Object)
 			}
 		case cache.Deleted:
 			if err := s.indexer.Delete(d.Object); err != nil {
 				return err
 			}
-			s.processor.distribute(deleteNotification{oldObj: d.Object})
+			s.OnDelete(d.Object)
 		}
 	}
 	return nil
 }
 
-type sharedProcessor struct {
-	listeners []*processorListener
-}
-
-func (p *sharedProcessor) distribute(obj interface{}) {
-	for _, listener := range p.listeners {
-		listener.add(obj)
+func (p *sharedIndexInformer) OnUpdate(oldObj, newObj interface{}) {
+	for _, handler := range p.handlers {
+		handler.OnUpdate(oldObj, newObj)
```

Comment:
If I'm reading this correctly, then if any handler gets stuck when called, it will prevent any further events from being received by any other handler.  One controller can crash all controllers.